### PR TITLE
Newparser

### PR DESCRIPTION
**Pull request by sbromberger**
This is a fix-all with a new parsing structure that takes advantage of Redis' CRLF line terminators. No more `readavailable` required, since we know when the Redis output is finished.

(In particular, this includes fixes for #1, #6, #7, and #8.)

### CODE REVIEW

File: src/commands.jl [hunk 1]
```diff
@@ -221,7 +221,7 @@ end
 @redisfunction "shutdown" String
 @redisfunction "shutdown" String option
 @redisfunction "slaveof" String host port
-@redisfunction "time" Array
+@redisfunction "_time" Array
```
Comment (sbromberger) [inline src/commands.jl]: I did this as a proof of concept to show that we can cast Redis output to Julia types if we need to. `time` seemed like an obvious first choice.

File: src/connection.jl [hunk 2]
```diff
@@ -98,9 +98,11 @@ function send_command(conn::RedisConnectionBase, command::String)
     write(conn.socket, command)
 end
 
-function execute_command(conn::RedisConnectionBase, command)
+function old_execute_command(conn::RedisConnectionBase, command)
```
Comment (sbromberger) [inline src/connection.jl]: Kept it here just in case.

File: src/commands.jl (hunk 1 above)
Comment (jkaye2012) [inline src/commands.jl]: Would you mind moving this part into another pr for discussion? There's probably value in doing this, but I'd like to keep things consistent in the mean time.

File: src/connection.jl (hunk 2 above)
Comment (jkaye2012) [inline src/connection.jl]: Let's just get rid of it. We can always pull it out of history if we need it

File: src/parser.jl [hunk 3]
```diff
@@ -1,82 +1,63 @@
...
-    end
+function parse_simple_string(l::AbstractString)
+    return l
 end
 
-# Simple strings replys extend to the first encountered CRLF
-function parse_simple_string_reply(reply, first_crlf)
-    ParsedReply(reply[2:first_crlf.start-1], first_crlf.stop)
+function parse_error(l::AbstractString)
+    error(l)
```
Comment (jkaye2012) [inline src/parser.jl]: Throw ServerException here?

File: src/parser.jl [hunk 4]
```diff
@@ -1,82 +1,63 @@
...
-# Integer replies are just ints followed by CRLF
-function parse_integer_reply(reply, first_crlf)
-    try
-        ParsedReply(parseint(reply[2:first_crlf.start-1]), first_crlf.stop)
-    catch
-        throw(ProtocolException(reply))
+function parse_bulk_string(s::TcpSocket, len::Int)
+    b = readbytes(s, len+2) # add crlf
+    if length(b) != len + 2
+        error("Bulk string read error: length(b) = $(length(b)), len = $len")
```
Comment (jkaye2012) [inline src/parser.jl]: Throw ProtocolException here?

File: src/commands.jl (hunk 1 above)
Comment (sbromberger) [inline src/commands.jl]: No worries. give me a few mins.

File: src/parser.jl (hunk 3 above)
Comment (sbromberger) [inline src/parser.jl]: Can do.

File: src/parser.jl [hunk 5]
```diff
@@ -1,82 +1,63 @@
...
+    if t == '+'
+        parse_simple_string(r)
+    elseif t == '-'
+        parse_error(r)
+    elseif t == ':'
+        parse_integer(r)
+    elseif t == '$'
+        len = parse_integer(r)
+        if len == -1
+            return ""
```
Comment (jkaye2012) [inline src/parser.jl]: Per the protocol spec, this should return null instead of an empty string

Comment (sbromberger) [inline src/parser.jl]: is Redis' `null` equivalent to Julia's `nothing`?

File: src/parser.jl [hunk 6]
```diff
@@ -1,82 +1,65 @@
...
-        ParsedReply(reply[first_crlf.stop+1:reply_end], reply_end+2)
-    catch
-        throw(ProtocolException(reply))
+function parse_integer(l::AbstractString)
+    return parse(Int, l)
+end
+
+function parse_array(s::TcpSocket, n::Int)
+    a = Vector{Any}()
+    for i = 1:n
```
Comment (jkaye2012) [inline src/parser.jl]: Should this not be 0:n? Also, we have to handle the null array response (n == -1)

File: src/parser.jl [hunk 7]
```diff
@@ -1,82 +1,65 @@
...
-        reply = reply[first_crlf.stop+1:end]
-        reply_length = first_crlf.stop
-        response = Any[]
-        for i=1:array_length
-            parsed_element = parse_reply(reply)
-            push!(response, parsed_element.response)
-            reply_length += parsed_element.reply_length
-            reply = reply[parsed_element.reply_length+1:end]
+function parseline(l::AbstractString, s::TcpSocket)
+    t = l[1]
```
Comment (jkaye2012) [inline src/parser.jl]: Consider renaming for clarity. Something like
t -> reply_type
r -> reply_token

File: src/parser.jl (hunk 5 above)
Comment (jkaye2012) [inline src/parser.jl]: Yup. The protocol spec defines `null` to be "whatever the nil value is in the client language"

File: src/parser.jl (hunk 6 above)
Comment (sbromberger) [inline src/parser.jl]: Redis doesn't have a concept of 0-basing - the integer passed after * is the number of elements in the array - and the null array response (n=0, per http://redis.io/topics/protocol#resp-arrays) is already handled since `parse_array` will return an empty vector..

Comment (jkaye2012) [inline src/parser.jl]: The null array is n == -1 (it's further down that spec). We still have to handle that case.

You're right about the indexing.

Comment (sbromberger) [inline src/parser.jl]: Ah, sorry - I got this mixed up with arrays. Let me fix that edge case.

**Comment by sbromberger** [comment]
Phew! I think that might do it, no?

**Comment by jkaye2012** [comment]
Looks pretty good to me. Let me pull it down quickly to run the tests, then I'll merge away.

Thanks for taking the time to help out on this!

**Comment by sbromberger** [comment]
Absolutely my pleasure. I have grand plans for Redis + [LightGraphs](https://github.com/JuliaGraphs/LightGraphs.jl).

**Comment by sbromberger** [comment]
Also, if you wouldn't mind tagging a new version, that way I can add this as a dependency in a project I've got going on.

**Comment by sbromberger** [comment]
PPS: I didn't make any changes to (or do any testing with) PubSub, so I don't know whether this impacts it.

**Comment by jkaye2012** [comment]
Sure, I'll tag the version once I merge it.

I think this change might break v0.3 compatibility though. When I run the tests, `get` fails with this message:

ERROR: type cannot be constructed
 in map at abstractarray.jl:1328
 in parse_bulk_string at /Users/jkaye/.julia/v0.3/Redis/src/parser.jl:26
 in parseline at /Users/jkaye/.julia/v0.3/Redis/src/parser.jl:58
 in execute_command at /Users/jkaye/.julia/v0.3/Redis/src/parser.jl:85
 in get at /Users/jkaye/.julia/v0.3/Redis/src/client.jl:70

Perhaps we can just require v0.4+ or something. Thoughts?

**Comment by sbromberger** [comment]
Ah. Yes. This is a @compat thing, I think. Could you test the latest commit to see whether that fixes things? (I don't have 0.3 to test with at the moment.)

**Comment by jkaye2012** [comment]
Getting closer. A new one with Vector{Any}

ERROR: type cannot be constructed
 in parse_array at /Users/jkaye/.julia/v0.3/Redis/src/parser.jl:35
 in parseline at /Users/jkaye/.julia/v0.3/Redis/src/parser.jl:58
 in execute_command at /Users/jkaye/.julia/v0.3/Redis/src/parser.jl:85
 in keys at /Users/jkaye/.julia/v0.3/Redis/src/client.jl:70

**Comment by sbromberger** [comment]
Yup, this is https://github.com/JuliaLang/Compat.jl/issues/105. Let me fix it. You'll want to change this once the Compat fix is in, though.

**Comment by sbromberger** [comment]
Can you retest with 0.3 now?

**Comment by jkaye2012** [comment]
Okay, pretty sure this will be the last problem we find.

Subscription is broken because of this line https://github.com/jkaye2012/Redis.jl/blob/master/src/client.jl#L49

Just have to plug the new parsing in there and everything should work. This would also be broken on v0.4; after it's fixed just run a local Redis server (with data that you don't care about - it will run `flushall`) and run `julia test/redis_tests.jl` to see if everything is good to go.

**Comment by jkaye2012** [comment]
Also, this highlights another problem, which is that the Travis-CI build isn't properly running the tests right now because runtests.jl doesn't have anything in it, so I will have to fix that at some point.

I'm pretty sure this commit actually breaks parser_tests.jl as well now that I think of it.

**Comment by sbromberger** [comment]
> Okay, pretty sure this will be the last problem we find.

Hah - you've just guaranteed that it won't be. :)

Thanks for the testing instructions. I'd been doing Pkg.test() without realizing that runtests was empty :)

**Comment by jkaye2012** [comment]
Yeah.. it's kinda goofy because redis_tests.jl is really the one you need to know that everything is working, but it's really an integration test and not a unit test.

**Comment by sbromberger** [comment]
OK, I'm having anxiety about pubsub here, since `SubscriptionMessage` relies on a `Reply` type that I'm not using.

**Comment by jkaye2012** [comment]
That reply was being used for the old parsing. If I'm not mistaken, subscription only uses the actual result, so you should be able to just pass the result of your new parsing into SubscriptionMessage and everything should be fine.

The way it works is that the subscription sends an array response to the client. Your parsing will turn that into an array of strings, which the SubscriptionMessage will parse for use by the frontend.

**Comment by sbromberger** [comment]
```
function SubscriptionMessage(reply)
        notification = reply.response
```

The new parser just returns an int, string, or array depending on the result of the command. There's no structure with a response attribute.

**Comment by jkaye2012** [comment]
Yeah, so you can remove the `notification = reply.response` and just use `reply` directly (which will be an array - the same that `reply.response` used to be)

**Comment by sbromberger** [comment]
Ah, I got it. pubsub is always an array, with the first element being the notification type. OK. stand by...

**Comment by sbromberger** [comment]
I kept this commit separate (unsquashed).

**Comment by sbromberger** [comment]
tests not working: specifically, 

```
@test x == ["hello, world!"]
```

**Comment by jkaye2012** [comment]
Missing conn.socket as second argument here I believe: https://github.com/sbromberger/Redis.jl/commit/cec86893a3ee63234d759d7d459326d992727c4c#diff-29a732d52d0661ba454242753a687486R50

**Comment by sbromberger** [comment]
Gah, of course. (it's my code; I should've caught that. Sorry!)

tests run now, though there are deprecation warnings, and I had a quick bugfix for an error in the script.

**Comment by jkaye2012** [comment]
Let's fix those in another commit. For now, this looks good on v0.3 and v0.4 so I'll go ahead and merge it once the build passes.

**Comment by sbromberger** [comment]
Just to be clear: I fixed the test script (by changing keys to Redis.keys) in this PR.

**Comment by sbromberger** [comment]
Thanks. Stand by for the fix for testing script.

**Comment by sbromberger** [comment]
See #11. BTW, thanks so much for your prompt review/feedback on this flurry of changes. :)
